src/x3d/x3dloadinternalstarling.pas

```diff
@@ -0,0 +1,611 @@
+    const TexCoordInterp: TCoordinateInterpolator2DNode);
+begin
+  CoordInterp.FdKeyValue.Items.AddRange(FCoordArray);
+  TexCoordInterp.FdKeyValue.Items.AddRange(FTexCoordArray);
+  { Repeat all keyValues, to avoid interpolating them smoothly between two keys }
+  CoordInterp.FdKeyValue.Items.AddRange(FCoordArray);
+  TexCoordInterp.FdKeyValue.Items.AddRange(FTexCoordArray);
+end;
+
+procedure TStarlingTextureAtlasLoader.AddAnimation(const FrameCount: Integer;
+  const TimeSensor: TTimeSensorNode;
+  const CoordInterp: TCoordinateInterpolatorNode;
+  const TexCoordInterp: TCoordinateInterpolator2DNode);
+var
+  I: Integer;
+  Key: Single;
+begin
+  { Set Cycle Interval becouse we know now frame count }
+  TimeSensor.CycleInterval := FrameCount / FFramesPerSecond;
+
+  { Generate list of keys. }
+  for I := 0 to FrameCount - 1 do
+  begin
+    Key := I / FrameCount;
+
+    CoordInterp.FdKey.Items.Add(Key);
+    TexCoordInterp.FdKey.Items.Add(Key);
+    if I > 0 then
+    begin
+      CoordInterp.FdKey.Items.Add(Key);
+      TexCoordInterp.FdKey.Items.Add(Key);
+    end;
+  end;
+
+  { This way, we have keys like
+    0 0.333 0.333 0.666 0.666 1
+    That is, all keys are repeated, except 0 and 1. }
+  CoordInterp.FdKey.Items.Add(1.0);
+  TexCoordInterp.FdKey.Items.Add(1.0);
+
+  { Add TimeSensor, CoordinateInterpolatorNode,
+    CoordinateInterpolator2DNode to Root node }
+  FRoot.AddChildren(TimeSensor);
+  FRoot.AddChildren(CoordInterp);
+  FRoot.AddChildren(TexCoordInterp);
+  AddRoutes(TimeSensor, CoordInterp, TexCoordInterp);
+end;
+
+procedure TStarlingTextureAtlasLoader.AddRoutes(
+  const TimeSensor: TTimeSensorNode;
+  const CoordInterp: TCoordinateInterpolatorNode;
+  const TexCoordInterp: TCoordinateInterpolator2DNode);
+var
+  R1, R2, R3, R4: TX3DRoute;
+begin
+  { Create routes. }
+  R1 := TX3DRoute.Create;
+  R2 := TX3DRoute.Create;
+  R3 := TX3DRoute.Create;
+  R4 := TX3DRoute.Create;
+  R1.SetSourceDirectly(TimeSensor.EventFraction_changed);
+  R1.SetDestinationDirectly(CoordInterp.EventSet_fraction);
+  R2.SetSourceDirectly(TimeSensor.EventFraction_changed);
+  R2.SetDestinationDirectly(TexCoordInterp.EventSet_fraction);
+  R3.SetSourceDirectly(CoordInterp.EventValue_changed);
+  R3.SetDestinationDirectly(FShapeCoord.FdPoint);
+  R4.SetSourceDirectly(TexCoordInterp.EventValue_changed);
+  R4.SetDestinationDirectly(FShapeTexCoord.FdPoint);
+  FRoot.AddRoute(R1);
+  FRoot.AddRoute(R2);
+  FRoot.AddRoute(R3);
+  FRoot.AddRoute(R4);
+end;
+
+function TStarlingTextureAtlasLoader.CheckAnimationNameAvailable(
+    const AnimationName: String): Boolean;
+begin
+  if FAnimationList.IndexOf(AnimationName) > -1 then
+  begin
+    WritelnWarning('Starling', 'Mixed animations tags (animation: %s) in "%s".', [AnimationName, FDisplayURL]);
+    Exit(false);
+  end;
+
+  FAnimationList.Add(AnimationName);
+  Result := true;
+end;
+
+
+constructor TStarlingTextureAtlasLoader.Create(const URL: String);
+begin
+  FURL := URL;
```

Comment:
Same as before, call `inherited Create` at the beginning of the constructor.

Also, place `Destroy` in the implementation right after `Create`, so move `Destroy` implementation a bit higher. We follow this convention in CGE: first `Create`, then `Destroy`, then the rest of methods in completely random order :)

Reply:
> Also, place Destroy in the implementation right after Create, so move Destroy implementation a bit higher. We follow this convention in CGE: first Create, then Destroy, then the rest of methods in completely random order :)

i thought this is just for the interface section, understand now, fixed.